### PR TITLE
lxd/image: set `Cached: false` for pre-existing cached image

### PR DESCRIPTION
**Pull request by gabrielmougard**
closes https://github.com/canonical/lxd/issues/13271

### CODE REVIEW

**Comment by tomponline** [comment]
test failures

**Comment by gabrielmougard** [comment]
@tomponline tests are passing now

File: lxd/daemon_images.go [hunk 1]
```diff
@@ -252,6 +254,23 @@ func ImageDownload(r *http.Request, s *state.State, op *operations.Operation, ar
 		ctxMap = logger.Ctx{"fingerprint": info.Fingerprint}
 		logger.Debug("Image already exists in the DB", ctxMap)
 
+		// If the image already exists, is cached and that it is
+		// requested to be downloaded from an explicit `image copy` operation, then disable its `cache` parameter
+		// so that it won't be candidate for auto removal.
+		if imgInfo.Cached && args.DownloadFromExplicitCopy && len(args.TargetAliasesNames) > 0 {
```
Comment (tomponline) [inline lxd/daemon_images.go:260]: Why does it need `len(args.TargetAliasesNames) > 0 ` part?

Comment (tomponline) [reply lxd/daemon_images.go:260]: can we do this as part of the transaction further down?

File: lxc/image.go [hunk 2]
```diff
@@ -256,7 +256,14 @@ func (c *cmdImageCopy) Run(cmd *cobra.Command, args []string) error {
 		imgInfo.Fingerprint = name
 	}
 
+	newAliases := make([]api.ImageAlias, len(c.flagAliases))
```
Comment (tomponline) [inline lxc/image.go:259]: Im not sure why we need this bit, please can you explain?

File: lxd/db/images.go [hunk 3]
```diff
@@ -773,6 +773,15 @@ func (c *ClusterTx) SetImageCachedAndLastUseDate(ctx context.Context, projectNam
 	return err
 }
 
+// UnsetImageCachedAndLastUseDate unsets the cached field of the image with the given fingerprint.
+func (c *ClusterTx) UnsetImageCachedAndLastUseDate(ctx context.Context, projectName string, fingerprint string, lastUsed time.Time) error {
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/db/images.go]: fixed

File: lxd/daemon_images.go (hunk 1 above)
Comment (gabrielmougard) [reply lxd/daemon_images.go]: A misunderstanding on my side.. There is not a real reason for why it is here. Removing it.

Comment (gabrielmougard) [reply lxd/daemon_images.go]: fixed

File: lxd/images.go [hunk 4]
```diff
@@ -415,18 +415,25 @@ func imgPostRemoteInfo(s *state.State, r *http.Request, req api.ImagesPost, op *
 		return nil, fmt.Errorf("must specify one of alias or fingerprint for init from image")
 	}
 
+	targetAliasesNames := make([]string, len(req.Aliases))
```
[the remark this answers is not in the document]
Comment (gabrielmougard) [reply lxd/images.go]: I don't think so. I removed it.

File: lxc/image.go (hunk 2 above)
Comment (gabrielmougard) [reply lxc/image.go]: An other misunderstanding... I thought we needed to check that the copy was creating a target alias, which is not necessarily the case. I removed it.

File: lxd/daemon_images.go [hunk 5]
```diff
@@ -31,20 +31,21 @@ import (
...
+	Alias                    string
+	Type                     string
+	SetCached                bool
+	PreferCached             bool
+	AutoUpdate               bool
+	Public                   bool
+	StoragePool              string
+	Budget                   int64
+	SourceProjectName        string
+	DownloadFromExplicitCopy bool
```
Comment (tomponline) [inline lxd/daemon_images.go:48]: As the struct is called `ImageDownload` the `Download` in `DownloadFromExplicitCopy` is redundant and stammers, e.g. `ImageDownload{}.DownloadFromExplicitCopy = true`

Lets just call it `UserRequested`.

Comment (tomponline) [reply lxd/daemon_images.go:48]: this aligns with https://github.com/canonical/lxd/blob/main/lxd/instance/instance_interface.go#L97

Comment (gabrielmougard) [reply lxd/daemon_images.go]: fixed

**Review by tomponline — APPROVED** [review]
LGTM thanks!

**Pull request merged**